### ISSUE TITLE
[ entropy_src, dv ] RFC: RNG samples dropped along non-seed boundaries

### ISSUE THREAD

**Issue by martin-lueker** [lowRISC/opentitan#10135]
The `entropy_src` IP has no real backpressure on the RNG.  It must continuously accept RNG inputs to continue monitoring them, even when the "final FIFO" is full.   Thus it must any drop unused seeds at the input of the final FIFO.   However in order to maintain verifiability data is dropped _only_ at the final fifo, and _only_ as complete seeds.

The DV environment works around this by keeping lists of complete candidate seeds and comparing output seeds to any possible matches.  Any candidate seed that must be skipped is assumed dropped and removed from the candidate list.   

This scheme is manageable enough for verification.  If, in addition, the entropy source were to arbitrarily drop 4-bit RNG inputs along non-seed boundaries, the necessary exploration space for verification could quickly become unmanageable. 

Unfortunately, _disabling_ and _re-enabling_ the entropy source has the effect of discarding an arbitrarily long sequence of RNG.  Currently the `entropy_src` does an effective reset as it is disabled:
- All the packers and FIFOs are cleared
- The counters within the health checks are cleared
- The SHA3 conditioner is purged of all data
- Once enabled again, the entire state is completely fresh.

If such a reset is done in the _middle_ of a seed calculation, the reset is then _completely equivalent_ to discarding the first portion of the seed.   The verification environment has no way of knowing how many RNG inputs are dropped in the process, and particularly when the conditioning is enabled, even missing one RNG input will change the output entirely.

***

Consider, for example, the following waveform: 
![image](https://user-images.githubusercontent.com/47870387/149880893-dadcfafc-4940-4388-8ecd-068ca423259f.png)

The signals should are taken from inside the adaptive proportion test.  The `entropy_src` has encountered several failing health tests and is in the alert state.  The virtual sequencer has caught the interrupt and is trying to bring the IP out of the alert state.   At Time D (red cursor, T=361ns) the enable register is set to MuBi-False and the `active_i` signal (light blue) strobes low.  This reset toggle actually happens twice in this waveform, but both pulses are able to reset the FSM and pull it from the alert state (FSM state not shown).

The simulation environment however is immediately unhappy.  At T=424 ns (end of sim) the sequencer checks the watermark register for this test and the scoreboard detects a discrepancy.   The last adaptive proportion test counted 185 1's, but the scoreboard thought that there should be eight more 1's somewhere, for 193.  Why the difference?

Both values are equally plausible.   (The average for this test is should be 192 with a standard deviation of &pm;9.8.)  However the score board does not know when the trigger was pulled to reset the DUT.

If we look at the usual heatlh check window boundaries (signal `window_wrap_pulse_i`, shown as an orange trace, with pulses demarcated by cursors at times A, B and C) we see that the typical health check window is &approx;40&mu;s.  The last disable pulse appears 20&mu;s into the last health check window.  About half the original data is thrown away but afterwards the DUT processes as usual, working for another 40&mu;s.   As currently written the scoreboard does not know about this shift, and uses the first 40&mu;s to calculate the test, but the DUT uses the latter 40&mu;s.   We've thrown away 20&mu;s of data and the scoreboard has no way of knowing which RNG samples the DUT decided to use.   

Based on previous testing we have reason to believe that DUT is performing the health check correctly on this later window.  However  once this shift takes place we have no way of accurately predicting what the DUT will do from here on out.

We need to find a way to constrain/predict this behavior.

***

In order to reproduce this result, pull the branch `issue_10135`  from my fork (`git@github.com:martin-lueker/opentitan.git`) and run the following  command (from `$OT_ROOT` on WD servers):
```
./util/dvsim/dvsim.py hw/ip/entropy_src/dv/entropy_src_sim_cfg.hjson -i entropy_src_rng -r 1 -t xcelium --job-prefix="nc run -C ncsim_interactive -Il" -w=shm -s=1 --run-modes set_verbosity_comp_a_uvm_debug set_verbosity_comp_b_uvm_debug
```

This last health check window starts at T=314.3&mu;s

**Comment by martin-lueker** [comment]
Note: this might be as simple as telling the scoreboard to discard all the RNG inputs when disabled.

**Comment by martin-lueker** [comment]
Fixed by #10243
